# Project: Profitable App Type Recommendation for a Company

## Project Summary
As data analysts, we are hired by a company that builds Android and iOS mobile apps. We have been tasked with helping a team of developers to make data-driven decisions with respect to the kind of apps they build. The goal is to increase revenue by attaracting more and more app users. We achieve this by finding mobile app profiles that are profitable for the App Store and Google Play markets.

## Company Profile
The company builds apps that are free to download and install, and the main source of revenue consists of in-app ads. This means that the revenue generated by any given app is mostly influenced by the number of users that use that particular app. Our goal for this project is to analyze data to help our developers understand what kinds of apps are likely to attract more users.


## Data Source: Opening and Exploration

As of February 2022, there were approximately `2.3 million iOS apps` available on the `App Store`, and `3.5 million Android apps` on `Google Play` as per [Statista](https://www.statista.com/statistics/276623/number-of-apps-available-in-leading-app-stores/). Collecting and analyzing data about all these apps (which is almost 6 million apps) requires a lot of time and resources. As this project is being carried out by a single person, so for the purpose of reducing the time and other resources required, we will consider only a subset of these applications. Rather than collecting data and creating a new dataset (which will be again a cubersome task for a single person), we will be using two datasets available online free of cost at [Kaggle-GooglePlay](https://www.kaggle.com/lava18/google-play-store-apps) and [Kaggle-AppStore](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps).

Let us start with opening and importing the datasets into our notebook.

In [1]:
from csv import reader

### Opening and Importing Google Play data set ###
opened_file = open('googleplaystore.csv')
read_file = reader(opened_file)
android_list = list(read_file)
android_header = android_list[0]
android_apps = android_list[1:]
### The App Store data set ###
opened_file = open('AppleStore.csv')
read_file = reader(opened_file)
ios_list = list(read_file)
ios_header = ios_list[0]
ios_apps = ios_list[1:]

Now that we have imported the contents of these datasets into two different lists, we will now create a function named `explore_data()`. This function will enable us to explore the data stored in lists created on a line-by-line basis

In [2]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line between rows
        
    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

### Exploring the Android Apps dataset

In [3]:
print(android_header)
print('\n')
explore_data(android_apps, 0, 3, True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows: 10841
Number of columns: 13


The Google Play data set has `10841` (rows) apps and `13` columns. Not all columns are useful for our purposes. The columns that seem to be useful for the purpose of our analysis are `'App', 'Category', 'Reviews', 'Installs', 'Type', 'Price', and 'Genres'`.


### Exploring the iOS Apps dataset

In [4]:
print(ios_header)
print('\n')
explore_data(ios_apps, 0, 3, True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 7197
Number of columns: 16


This dataset has `7197 iOS apps` (rows) and `16 column`. The columns that seem interesting in this dataset are: `'track_name', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', and 'prime_genre'`.

## Enriching the Dataset (Pre-Processing)

### Detecting and Deleting Wrong Data
From the discussions section of Google Play dataset at [Kaggle-GooglePlay](https://www.kaggle.com/lava18/google-play-store-apps/discussion), we can see that one of the [discussions](https://www.kaggle.com/lava18/google-play-store-apps/discussion/66015) outlines an error for row 10472. Let's print this row and try to find what is causing the error by comparing it against the header and another row that is correct.

In [5]:
print(android_apps[10472])  # incorrect row
print('\n')
print(android_header)  # header
print('\n')
print(android_apps[0])      # correct row

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


As we can see from the output above, the row 10472 corresponds to the app named "Life Made WI-Fi Touchscreen Photo Frame", and its rating is 19. As per the information provided in the [discussions](https://www.kaggle.com/lava18/google-play-store-apps/discussion/66015) section, this is clearly wrong, because the maximum possible rating for a Google Play app is 5. SO,this problem is caused by a missing value in the 'Category' column. To remove this error, we'll delete this row.

In [6]:
print("Total number of records (rows) before deletion:", len(android_apps))
#del android_apps[10472]  # Avoid re-running this piece of code, as it will result in loss of good data
print("Total number of records (rows) after deletion:", len(android_apps))

Total number of records (rows) before deletion: 10841
Total number of records (rows) after deletion: 10841


### Detecting Duplicate Records (rows)

To manually detect and remove the duplicate records (rows) from these datasets is cubersome. So, instead of going through datasets manually, we loaded both datasets ([Kaggle-GooglePlay](https://www.kaggle.com/lava18/google-play-store-apps) and [Kaggle-AppStore](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps)) in `Microsoft Excel` and run the `"Remove Duplicates"` command from the `"Table Design"` menu. 

#### Detecting the duplicates in the Google Play Dataset

From Excel analysis of the Google Play dataset, we found that there were `1181` duplicate entries in the `App` column, which contains the name of different apps. There were `9659` unique records(rows) in this dataset. To confirm our findings, we analysed the "App" column of the dataset, which contains the Android app names. The following code produced the same results.

In [7]:
duplicate_android_apps = []
unique_android_apps = []

for app_row in android_apps:
    app_name = app_row[0]
    if app_name in unique_android_apps:
        duplicate_android_apps.append(app_name)
    else:
        unique_android_apps.append(app_name)
    
print('Number of duplicate apps in Google Play dataset:', len(duplicate_android_apps))
print('\n')
print('Examples of duplicate apps in Google Play dataset:', duplicate_android_apps[:15])

Number of duplicate apps in Google Play dataset: 1181


Examples of duplicate apps in Google Play dataset: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software']


#### Detecting the duplicates in the iOS Apps Dataset
`No duplicates` were found in the `iOS apps` dataset.

After confirming that there are duplicate records(rows) in the `Google Play` dataset, we will now proceed to removing the duplicate entries.
But, before starting deleting the duplicate rows, we need to formulate a criteria for selecting the rows to delete. We could go with randomly deleting duplicate rows, but it is not considered a good practice and also is not efficient. We will like to retain the row which is most useful for our purpose. For this reason, we will have a look at the different fields(columns) of the duplicate entiries and try to figure out what field value differntiates these duplicate entries. 

Upon examining the different duplicate rows, we observe that the main difference of column values happens on the fourth column of each row, which corresponds to the number of reviews. The different numbers show that the data was collected at different times. We will use this column to build a criterion for determining the deletion or retention of a particular duplicate row. We will keep the rows that have the highest number of reviews because the higher the number of reviews, the more reliable the ratings.

### Deleting Duplicate Data (rows)
In order to remove the duplicate records (rows), we proceed as follows:
1. Create a dictionary where each key is a unique app name, and the value is the highest number of reviews of that app
2. Use the dictionary to create a new data set, which will have only one entry per app (and we only select the apps with the highest number of reviews)

**Building a dictionary**

In [8]:
max_num_reviews = {}
number_reviews = 0

for app_data in android_apps:
    name_app = app_data[0]
    temp_str = app_data[3]
    if temp_str[-1] == "M":
        temp_str = temp_str[:-1]
    number_reviews = temp_str
    
    if name_app in max_num_reviews and max_num_reviews[name_app] < number_reviews:
        max_num_reviews[name_app] = number_reviews
        
    elif name_app not in max_num_reviews:
        max_num_reviews[name_app] = number_reviews

Previously, in the `"Detecting the duplicates in the Google Play Dataset"` section, we found that there are `1,181 duplicate records` or rows, so the length of list containing unique apps should be equal to `9659`.

In [9]:
print('Number of unique rows/records:', len(max_num_reviews))

Number of unique rows/records: 9660


#### Removing Duplicate Records (rows)
In order to remove the duplicate rows/records from the data, we will use the `"max_reviews"` dictionary we created in a previous cell. As per the criterion laid down earlier, among the duplicate records/rows, we will keep the rows with highest number of reviews for same app.  

To create a dataset with unique rows only, we will proceed as follows:
1. We will start by initializing two empty lists, `android_unique_ds` and `already_added_app`.
2. We will loop through the `android-apps` data set, and for every iteration, we will do the following steps:
    i. isolate the name of the app and the number of reviews.
    ii. add the current row to the `android_unique_ds` list, and the app name to the `already_added_app` list if:
        * number of reviews of the current app matches the number of reviews of that app stored in the `max_reviews` dictionary; and
        * name of the app is not already in the `already_added_app` list. 
*{We need to add this second condition to account for those cases where the highest number of reviews of a duplicate app is the same for more than one entry. If we just check for max_reviews[app_name] == n_reviews, we'll still end up with duplicate entries for some apps.}*

In [10]:
android_unique_ds = []
already_added_app = []

for appl_row in android_apps:
    appl_name = appl_row[0]
    number_reviews = appl_row[3]
    
    if number_reviews[-1] == "M":
        number_reviews = number_reviews[:-1]
       
    if (max_num_reviews[appl_name] == number_reviews) and (appl_name not in already_added_app):
        android_unique_ds.append(appl_row)
        already_added_app.append(appl_name)

Now, let us check and confirm that the number of rows in newly constructd dataset `android_unique_ds` is exactly `9,659`.

In [11]:
explore_data(android_unique_ds, 0, 5, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']


['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+', 'Free', '0', 'Everyone', 'Art & Design', 'March 26, 2017', '1.0', '2.3 and up']


Number of rows: 9660
Number of columns: 13


### Removing Non-English Apps
While analysing the original datasets (i.e. googleplaystore and AppleStore) in Excel, we found that some values in the `track_name` field (which represents the app name) are written in non-English languages. This means those apps are targetted more towards non-English speaking users. 

In [12]:
print(ios_apps[6609][1])
print(ios_apps[7174][1])

print(android_unique_ds[4412][0])
print(android_unique_ds[7940][0])

最後の英単語学習！マジタン
２ちゃんねる for iPhone
Wowkwis aq Ka'qaquj
PHARMAGUIDE (DZ)


We will remove all the records/rows associated with an app whose name is not in English. To remove such rows, we will iterate over the dataset checking the `track_name` field for app names which contain characters, special chatracters, numbers and symbols which are not included in the English language. The set of characters, special chatracters, numbers and symbols associated with English language fall in the `ASCII` code range of `0 - 127`. We will use these ASCII codes to check whether a value for app name in the `track_name`field contains non-ASCII characters. 
For this purpose, we will write a fuction named `check_english()`. For finding ASCII code of each character, we will use the in-built called `ord()`. The `check_english()` function takes as input the name of an app and returns a `True` or `False`.

In [13]:
def check_english(input_str):
    
    for char in input_str:
        if ord(char) > 127:
            return False
    
    return True

print(ios_apps[10][1], ':', check_english(ios_apps[10][1]))
print(ios_apps[6609][1], ':', check_english(ios_apps[6609][1]))

print(android_unique_ds[5][0], ':', check_english(android_unique_ds[5][0]))
print(android_unique_ds[4412][0], ':', check_english(android_unique_ds[4412][0]))

Subway Surfers : True
最後の英単語学習！マジタン : False
Smoke Effect Photo Maker - Smoke Editor : True
Wowkwis aq Ka'qaquj : True


The `check_english()` function would work fine, if there were only plain English and non-English app names. But, as observed during the dataset analysis, some apps use emojis in their names. If we process our dataset with this function, we may loose some important apps data. For example:

In [14]:
print(ios_apps[46][1], ':', check_english(ios_apps[46][1]))
print(ios_apps[164][1], ':', check_english(ios_apps[164][1]))

print(android_apps[139][0], ':', check_english(android_apps[139][0]))
print(android_apps[418][0], ':', check_english(android_apps[418][0]))

▻Sudoku : False
▻Solitaire : False
Wattpad 📖 Free Books : False
Dolphin Browser - Fast, Private & Adblock🐬 : False


The above example demonstartes how we may loose some useful data if we process our datasets with the `check_english()` function. To avoid this loss of data, we will only remove those apps that have more than three non-ASCII characters in the `track_name`field i.e. app name. For this purpose, we will modify `check_english()` function and also change its name to `check_english_emoji()`, to reflect the modifications. 

In [15]:
def check_english_emoji(input_str):
    non_ascii = 0
    
    for char in input_str:
        if ord(char) > 127:
            non_ascii += 1
    
    if non_ascii > 3:
        return False
    else:
        return True

Now, let us call the `check_english_emoji()` function and give as input the same data we gave to `check_english()` function in previous cell i.e, non-English app names and app names with emojis.

**Input with Non-English App Names**

In [16]:
print(ios_apps[10][1], ':', check_english_emoji(ios_apps[10][1]))
print(ios_apps[6609][1], ':', check_english_emoji(ios_apps[6609][1]))

print(android_unique_ds[5][0], ':', check_english_emoji(android_unique_ds[5][0]))
print(android_unique_ds[4412][0], ':', check_english_emoji(android_unique_ds[4412][0]))

Subway Surfers : True
最後の英単語学習！マジタン : False
Smoke Effect Photo Maker - Smoke Editor : True
Wowkwis aq Ka'qaquj : True


**Input with Upto 3 Non-ASCII Characters**

In [17]:
print(ios_apps[46][1], ':', check_english_emoji(ios_apps[46][1]))
print(ios_apps[164][1], ':', check_english_emoji(ios_apps[164][1]))

print(android_apps[139][0], ':', check_english_emoji(android_apps[139][0]))
print(android_apps[418][0], ':', check_english_emoji(android_apps[418][0]))

▻Sudoku : True
▻Solitaire : True
Wattpad 📖 Free Books : True
Dolphin Browser - Fast, Private & Adblock🐬 : True


**Using "check_english_emoji()" function to remove non-English app names from both the datasets**

We will start with declaring one empty list for each dataset i.e, `android_english_app` and `ios_english_app`. This empty list will be used to store the list of apps that pass our English app name constraint. Next, we will iterate through `android_unique_ds` dataset and `ios_apps` dataset, checking for app name validity.

In [18]:
android_english_app = []
ios_english_app = []

for app_row in android_unique_ds:
    app_name = app_row[0]
    if check_english_emoji(app_name):
        android_english_app.append(app_row)
        
for app_row in ios_apps:
    app_name = app_row[1]
    if check_english_emoji(app_name):
        ios_english_app.append(app_row)
        
explore_data(android_english_app, 0, 3, True)
print('\n')
explore_data(ios_english_app, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9615
Number of columns: 13


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 

### Creating a New Dataset By Extracting Free Apps Data

We would like to recall here that the company we are working for builds free to use apps only. Also, that the company's revenue is directly dependent on the revenue generated from the in-app ads. The datasets we have contain data about both free and non-free apps. As non-free apps are not usually dependent on in-app ad revenue generation, they have a different marketing strateggy. SO, now onwards, our focus will be on the data related to free apps. To eliminate the burden of unnecessary processing of data related to non-free apps, we will create a separate lists from free apps related data in both our original datasets.
For this purpose, we will use `android_english_app` and `ios_english_app` datasets. These datasets are the ones that are free from corrupt column values, duplicates and non-English app names. 

In [19]:
android_final_ds = []
ios_final_ds = []

for app_row in android_english_app:
    price = app_row[7]
    if price == '0':
        android_final_ds.append(app_row)
        
for app_row in ios_english_app:
    price = app_row[4]
    if price == '0.0':
        ios_final_ds.append(app_row)
        
print('Number of free Android apps: ', len(android_final_ds)) # to check the final number of records in Android

print('Number of free iOS apps: ', len(ios_final_ds)) # to check the final number of records in iOS

Number of free Android apps:  8862
Number of free iOS apps:  3222


## Identifying Most Common App Profiles

Because our end goal is to build an app and add that app on both Google Play and the App Store, we need to find app profiles that are successful in both markets. For instance, a profile that works well for both markets might be a productivity app that makes use of gamification. Next, step is to find the type of data fields that is present in both datasets. Out of this data, we will need to identify what data fields will help us to identify the type of apps that are the most successful ones on both the platforms. Based on this analysis, we can advise the company as to which type of app will help the company to increase its in-app ads based revenue generation.

From the analysis we did on both the data sets, we got a sense of the most common genres for each market. Based on this analysis, we will start with building a frequency table for the `prime_genre` column of the `App Store` data set, and the `Genres` and `Category` columns of the `Google Play` data set.

To minimize risks and overhead, our validation strategy for an app idea has three steps:
- Build a minimal Android version of the app, and add it to Google Play.
- If the app has a good response from users, we develop it further.
- If the app is profitable after six months, we build an iOS version of the app and add it to the App Store. 

### Generate Frequency Tables That Show Percentages

We will write a function for that generates frequency tables. We know that dictionaries don't have order, and it will be very difficult to analyze the frequency tables. We will write a second function that will display the entries in the frequency table in descending order. We will use the built-in `sorted()` function for this purpose. This function takes in an iterable data type (like a list, dictionary, tuple, etc.), and returns a list of the elements of that iterable sorted in ascending or descending order (the `reverse` parameter controls whether the order is ascending or descending).

In [20]:
def freq_table(dataset, index):
    table = {}
    total = 0
    
    for row in dataset:
        total += 1
        value = row[index]
        if value in table:
            table[value] += 1
        else:
            table[value] = 1
    
    table_percentages = {}
    for key in table:
        percentage = (table[key] / total) * 100
        table_percentages[key] = percentage 
    
    return table_percentages


def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)
        
    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

### Display Frequency Tables

To display frequency tables for different columns in datasets, we will call the `display_table()` function defined above multiple times. Each time we will pass different column as an argument.

**Display Frequency Table for `prime_genre` Column in the `AppleStore` Dataset**

In [21]:
display_table(ios_final_ds, -5)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


After looking at the frequency table above, we can see that among the free English apps, more than a half `(58.16%)` are `Games`. `Entertainment` apps are a little more than `7%`, followed by `Education` apps, which are over `6.6%`. Only `1.3%` of the apps are designed for `Shopping`, in iOS dataset.

The results in the frequency table above establish a general impression that the majority of `Free English Apps` hosted on the `Apple App Store` are designed for fun (games, entertainment, photo and video, social networking, sports, music, etc.), while apps with practical purposes such as education, shopping, utilities, productivity, lifestyle, etc. are more rare. However, the fact that fun apps are the most prominent doesn't imply that they also have the greatest number of users — the demand might not be the same as the offer.

***Display Frequency Tables for `Genres` and `Category` Columns of the `Google Play` Dataset***

Let us first generate and analyse frequency table for the `Category` column.

In [22]:
display_table(android_final_ds, 1)

FAMILY : 18.934777702550214
GAME : 9.693071541412774
TOOLS : 8.451816745655607
BUSINESS : 4.5926427443015125
LIFESTYLE : 3.9043105393816293
PRODUCTIVITY : 3.8930264048747465
FINANCE : 3.7011961182577298
MEDICAL : 3.5206499661475967
SPORTS : 3.39652448657188
PERSONALIZATION : 3.3175355450236967
COMMUNICATION : 3.238546603475513
HEALTH_AND_FITNESS : 3.080568720379147
PHOTOGRAPHY : 2.945159106296547
NEWS_AND_MAGAZINES : 2.798465357707064
SOCIAL : 2.663055743624464
TRAVEL_AND_LOCAL : 2.335815842924848
SHOPPING : 2.2455427668697814
BOOKS_AND_REFERENCE : 2.143985556307831
DATING : 1.8618821936357481
VIDEO_PLAYERS : 1.7941773865944481
MAPS_AND_NAVIGATION : 1.399232678853532
FOOD_AND_DRINK : 1.2412547957571656
EDUCATION : 1.1735499887158656
ENTERTAINMENT : 0.9591514330850823
LIBRARIES_AND_DEMO : 0.9365831640713158
AUTO_AND_VEHICLES : 0.9252990295644324
HOUSE_AND_HOME : 0.8237418190024826
WEATHER : 0.8011735499887158
EVENTS : 0.7109004739336493
PARENTING : 0.6544798013992327
ART_AND_DESIGN : 0.

The results in above frequency table based on `Category` column in the `Google Play Dataset` are significantly different from the one generated for `prime_genre` column in the `AppleStore Dataset`: we can see that the apps designed for fun are the most prominent ones, and a large number of apps are designed for practical purposes such as family, tools, business, lifestyle, productivity, etc. What can be deduced from the above results is that practical apps are more popular on Google Play compared to App Store.

Let us now generate and analyse frequency table for the `Genres` column.

In [25]:
display_table(android_final_ds, -4)

Tools : 8.440532611148726
Entertainment : 6.070864364703228
Education : 5.348679756262695
Business : 4.5926427443015125
Productivity : 3.8930264048747465
Lifestyle : 3.8930264048747465
Finance : 3.7011961182577298
Medical : 3.5206499661475967
Sports : 3.4642292936131795
Personalization : 3.3175355450236967
Communication : 3.238546603475513
Action : 3.1031369893929135
Health & Fitness : 3.080568720379147
Photography : 2.945159106296547
News & Magazines : 2.798465357707064
Social : 2.663055743624464
Travel & Local : 2.324531708417964
Shopping : 2.2455427668697814
Books & Reference : 2.143985556307831
Simulation : 2.0424283457458814
Dating : 1.8618821936357481
Arcade : 1.8505980591288649
Video Players & Editors : 1.7716091175806816
Casual : 1.7490408485669149
Maps & Navigation : 1.399232678853532
Food & Drink : 1.2412547957571656
Puzzle : 1.128413450688332
Racing : 0.9930038366057323
Role Playing : 0.9365831640713158
Libraries & Demo : 0.9365831640713158
Auto & Vehicles : 0.92529902956443

There are no obvious differences between frequency tables generated for the `Genres` and the `Category` columns. The most obvious thing that one can notice is that the `Genres` column has many more number of categories. Now onwards, we'll only work with the `Category` column to get an overall picture of what is good for the company rather than digging deeper by considering all different dataset columns.

So far, we noticed that with regard to free apps, the `App Store` is dominated by apps designed for fun, while `Google Play` maintians a more balanced blend of both practical and for-fun apps. Now, let's move forward and try to get an idea about the kind of apps that have most users.

### Finding Most Popular Apps by Analysing `Genre` Column in the App Store Dataset

In order to decide, what genres of apps have the most number of users, we will calculate the average number of installs for each app genre. For the Google Play data set, we can find this information in the `Installs` column, but for the App Store data set this information is missing. As a workaround, we'll take the total number of user ratings as a proxy, which we can find in the `rating_count_tot` app.

Below, we calculate the average number of user ratings per app genre on the `App Store`:

In [26]:
genres_ios = freq_table(ios_final_ds, -5)

for genre in genres_ios:
    total = 0
    len_genre = 0
    for app in ios_final_ds:
        genre_app = app[-5]
        if genre_app == genre:            
            n_ratings = float(app[5])
            total += n_ratings
            len_genre += 1
    avg_n_ratings = total / len_genre
    print(genre, ':', avg_n_ratings)

Social Networking : 71548.34905660378
Photo & Video : 28441.54375
Games : 22788.6696905016
Music : 57326.530303030304
Reference : 74942.11111111111
Health & Fitness : 23298.015384615384
Weather : 52279.892857142855
Utilities : 18684.456790123455
Travel : 28243.8
Shopping : 26919.690476190477
News : 21248.023255813954
Navigation : 86090.33333333333
Lifestyle : 16485.764705882353
Entertainment : 14029.830708661417
Food & Drink : 33333.92307692308
Sports : 23008.898550724636
Book : 39758.5
Finance : 31467.944444444445
Education : 7003.983050847458
Productivity : 21028.410714285714
Business : 7491.117647058823
Catalogs : 4004.0
Medical : 612.0


As we can see, `navigation apps` have the highest number of user reviews, but this figure is heavily influenced by `Waze` and `Google Maps` apps, which have close to half a million user reviews together:

In [28]:
for app in ios_final_ds:
    if app[-5] == 'Navigation':
        print(app[1], ':', app[5]) # print name and number of ratings

Waze - GPS Navigation, Maps & Real-time Traffic : 345046
Google Maps - Navigation & Transit : 154911
Geocaching® : 12811
CoPilot GPS – Car Navigation & Offline Maps : 3582
ImmobilienScout24: Real Estate Search in Germany : 187
Railway Route Search : 5


The same pattern applies to `social networking` apps, where the average number is heavily influenced by a few giants like `Facebook, Pinterest, Skype,` etc. Same applies to `music apps`, where a few big players like `Pandora, Spotify,` and `Shazam` heavily influence the average number.

The objective of this project is to find popular genres of apps that can help company to decide which app can help them maximise the revenue, but `navigation, social networking or music` apps might seem more popular than they really are. The average number of ratings seem to be skewed by very few apps which have hundreds of thousands of user ratings, while the other apps may struggle to get past the 10,000 threshold.

`Reference` apps have 74,942 user ratings on average, but it's actually the `Bible and Dictionary.com` which skew up the average rating:

In [29]:
for app in ios_final_ds:
    if app[-5] == 'Reference':
        print(app[1], ':', app[5])

Bible : 985920
Dictionary.com Dictionary & Thesaurus : 200047
Dictionary.com Dictionary & Thesaurus for iPad : 54175
Google Translate : 26786
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran : 18418
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition : 17588
Merriam-Webster Dictionary : 16849
Night Sky : 12122
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE) : 8535
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools : 4693
GUNS MODS for Minecraft PC Edition - Mods Tools : 1497
Guides for Pokémon GO - Pokemon GO News and Cheats : 826
WWDC : 762
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free : 718
VPN Express : 14
Real Bike Traffic Rider Virtual Reality Glasses : 8
教えて!goo : 0
Jishokun-Japanese English Dictionary & Translator : 0


From what we have observed from the analysis of these datasets so far, the company could build another popular book app, which will provide some additional features (such as bookmarking, in-app search, comments etc.) besides the raw version of the book. This might also include daily quotes from the book, an audio version of the book, quizzes about the book, etc. On top of that, they can embed a dictionary within the app, so users don't need to exit the app to look up words in an external app.

This idea seems to fit well with the fact that the App Store is dominated by for-fun apps. This suggests the market might be a bit saturated with for-fun apps, which means a practical app might have more of a chance to stand out among the huge number of apps on the App Store.

Other genres that seem popular include weather, book, food and drink, or finance. The book genre seem to overlap a bit with the app idea we described above, but the other genres don't seem too interesting to us:

- Weather apps — people generally don't spend too much time in-app, and the chances of making profit from in-app adds are low. Also, getting reliable live weather data may require us to connect our apps to non-free APIs.

- Food and drink — examples here include Starbucks, Dunkin' Donuts, McDonald's, etc. So making a popular food and drink app requires actual cooking and a delivery service, which is outside the scope of our company.

- Finance apps — these apps involve banking, paying bills, money transfer, etc. Building a finance app requires domain knowledge, and we don't want to hire a finance expert just to build an app.

Now let's analyze the Google Play market a bit.

### Finding Most Popular Apps by Analysing Genres Column in the Google Play Dataset

For the Google Play market, we actually have data about the number of installs, so we are be able to get a clearer picture about genre popularity. However, the install numbers don't seem precise enough — we can see that most values are open-ended (100+, 1,000+, 5,000+, etc.):

In [30]:
display_table(android_final_ds, 5) # the Installs columns

1,000,000+ : 15.741367637102236
100,000+ : 11.554953735048521
10,000,000+ : 10.516813360415256
10,000+ : 10.200857594222523
1,000+ : 8.395396073121193
100+ : 6.917174452719477
5,000,000+ : 6.838185511171294
500,000+ : 5.574362446400361
50,000+ : 4.773188896411646
5,000+ : 4.513653802753328
10+ : 3.5432182351613632
500+ : 3.2498307379823967
50,000,000+ : 2.2906793048973144
100,000,000+ : 2.1214172872940646
50+ : 1.9183028661701647
5+ : 0.7898894154818324
1+ : 0.5077860528097494
500,000,000+ : 0.2708192281651997
1,000,000,000+ : 0.22568269013766643
0+ : 0.045136538027533285
0 : 0.011284134506883321


To perform computations, we'll need to convert each install number to `float` — this means that we need to remove the commas and the plus characters, otherwise the conversion will fail and raise an error. We'll do this directly in the loop below, where we also compute the average number of installs for each genre (category).

In [31]:
categories_android = freq_table(android_final_ds, 1)

for category in categories_android:
    total = 0
    len_category = 0
    for app in android_final_ds:
        category_app = app[1]
        if category_app == category:            
            n_installs = app[5]
            n_installs = n_installs.replace(',', '')
            n_installs = n_installs.replace('+', '')
            total += float(n_installs)
            len_category += 1
    avg_n_installs = total / len_category
    print(category, ':', avg_n_installs)

ART_AND_DESIGN : 1986335.0877192982
AUTO_AND_VEHICLES : 647317.8170731707
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 8767811.894736841
BUSINESS : 1712290.1474201474
COMICS : 817657.2727272727
COMMUNICATION : 38456119.167247385
DATING : 854028.8303030303
EDUCATION : 1820673.076923077
ENTERTAINMENT : 11640705.88235294
EVENTS : 253542.22222222222
FINANCE : 1387692.475609756
FOOD_AND_DRINK : 1924897.7363636363
HEALTH_AND_FITNESS : 4188821.9853479853
HOUSE_AND_HOME : 1331540.5616438356
LIBRARIES_AND_DEMO : 638503.734939759
LIFESTYLE : 1437816.2687861272
GAME : 15560965.599534342
FAMILY : 3694276.334922527
MEDICAL : 120616.48717948717
SOCIAL : 23253652.127118643
SHOPPING : 7036877.311557789
PHOTOGRAPHY : 17805627.643678162
SPORTS : 3638640.1428571427
TRAVEL_AND_LOCAL : 13984077.710144928
TOOLS : 10682301.033377837
PERSONALIZATION : 5201482.6122448975
PRODUCTIVITY : 16787331.344927534
PARENTING : 542603.6206896552
WEATHER : 5074486.197183099
VIDEO_PLAYERS : 24727872.452830188
NEWS_AND_

On average, `Communication apps` have the most installs i.e. `38,456,119`. This number is heavily skewed up by a few apps that have over one billion installs such as `WhatsApp, Facebook Messenger, Skype, Google Chrome, Gmail, and Hangouts`, and a few others with over 100 and 500 million installs. Let us have a look at install statistics of different apps in this group:

In [32]:
for app in android_final_ds:
    if app[1] == 'COMMUNICATION' and (app[5] == '1,000,000,000+'
                                      or app[5] == '500,000,000+'
                                      or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

WhatsApp Messenger : 1,000,000,000+
imo beta free calls and text : 100,000,000+
Android Messages : 100,000,000+
Google Duo - High Quality Video Calls : 500,000,000+
Messenger – Text and Video Chat for Free : 1,000,000,000+
imo free video calls and chat : 500,000,000+
Skype - free IM & video calls : 1,000,000,000+
Who : 100,000,000+
GO SMS Pro - Messenger, Free Themes, Emoji : 100,000,000+
LINE: Free Calls & Messages : 500,000,000+
Google Chrome: Fast & Secure : 1,000,000,000+
Firefox Browser fast & private : 100,000,000+
UC Browser - Fast Download Private & Secure : 500,000,000+
Gmail : 1,000,000,000+
Hangouts : 1,000,000,000+
Messenger Lite: Free Calls & Messages : 100,000,000+
Kik : 100,000,000+
KakaoTalk: Free Calls & Text : 100,000,000+
Opera Mini - fast web browser : 100,000,000+
Opera Browser: Fast and Secure : 100,000,000+
Telegram : 100,000,000+
Truecaller: Caller ID, SMS spam blocking & Dialer : 100,000,000+
UC Browser Mini -Tiny Fast Private & Secure : 100,000,000+
Viber Mess

If the communication apps with over 100 million installs are removed from the calculation, the average install count will be reduced almost bya factor of ten. Let us remove such apps and see what do we get.

In [33]:
under_100_m = []

for app in android_final_ds:
    n_installs = app[5]
    n_installs = n_installs.replace(',', '')
    n_installs = n_installs.replace('+', '')
    if (app[1] == 'COMMUNICATION') and (float(n_installs) < 100000000):
        under_100_m.append(float(n_installs))
        
sum(under_100_m) / len(under_100_m)

3603485.3884615386

The `video players` category, which is the runner-up with 24,727,872 installs,shows the same pattern. The market is dominated by apps like `Youtube, Google Play Movies & TV, or MX Player`. The pattern is also repeated for other app categories such as `social apps`, `photography apps`, and `productivity apps`.

These app genres might seem more popular than they really are. In reality, the reason for high install counts for these genres are due to a few popular apps.

The `game` genre is pretty popular, but as we observed earlier, this part of the market is saturated with a lot of free apps already in the market, so we'd like to come up with a different app recommendation if possible.

The `books and reference genre` looks fairly popular as well, with an average number of installs of `8,767,811`. It's interesting to explore this in more depth, since we found this genre has some potential to work well on the `App Store`, and our aim is to recommend an app genre that shows potential for being profitable on both the `App Store` and `Google Play`. 

Let's take a look at some of the apps from this genre and their number of installs:

In [34]:
for app in android_final_ds:
    if app[1] == 'BOOKS_AND_REFERENCE':
        print(app[0], ':', app[5])

E-Book Read - Read Book for free : 50,000+
Download free book with green book : 100,000+
Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Free Panda Radio Music : 100,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
English Grammar Complete Handbook : 500,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
Google Play Books : 1,000,000,000+
AlReader -any text book reader : 5,000,000+
Offline English Dictionary : 100,000+
Offline: English to Tagalog Dictionary : 500,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
Recipes of Prophetic Medicine for free : 500,000+
ReadEra – free ebook reader : 1,000,000+
Anonymous caller detection : 10,000+
Ebook Reader : 5,000,000+
Litnet - E-books : 100,000+
Read books online : 5,000,000+
English to Urdu Dictionary : 500,000+
eBoox: book reader fb2 epub zip : 1,000,000+
English Persian Dictionary : 500,000+
Flybook : 500,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
E

The `book and reference` genre includes a variety of apps: software for processing and reading ebooks, various collections of libraries, dictionaries, tutorials on programming or languages, etc. It seems there's still a small number of extremely popular apps that skew the average:

In [35]:
for app in android_final_ds:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000,000+'
                                            or app[5] == '500,000,000+'
                                            or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

Google Play Books : 1,000,000,000+
Bible : 100,000,000+
Amazon Kindle : 100,000,000+
Wattpad 📖 Free Books : 100,000,000+
Audiobooks from Audible : 100,000,000+


However, it looks like there are only a few very popular apps, so this market still shows potential. Let's try to get some app ideas based on the kind of apps that are somewhere in the middle in terms of popularity (between 1,000,000 and 100,000,000 downloads):

In [36]:
for app in android_final_ds:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000+'
                                            or app[5] == '5,000,000+'
                                            or app[5] == '10,000,000+'
                                            or app[5] == '50,000,000+'):
        print(app[0], ':', app[5])

Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
AlReader -any text book reader : 5,000,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
ReadEra – free ebook reader : 1,000,000+
Ebook Reader : 5,000,000+
Read books online : 5,000,000+
eBoox: book reader fb2 epub zip : 1,000,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
Moon+ Reader : 10,000,000+
English-Myanmar Dictionary : 1,000,000+
Golden Dictionary (EN-AR) : 1,000,000+
All Language Translator Free : 1,000,000+
Aldiko Book Reader : 10,000,000+
Dictionary - WordWeb : 5,000,000+
50000 Free eBooks & Free AudioBooks : 5,000,000+
Al-Quran (Free) : 10,000,000+
Al Quran Indonesia : 10,000,000+
Al'Quran Bahasa Indonesia : 10,000,000+
Al Quran Al karim : 1,000,000+
Al Quran : EAlim - Translations & MP3 Offline : 5,000,000+
Koran Read &MP3 30 Juz Offline : 1,000,000+
H

This niche seems to be dominated by software for processing and reading ebooks, as well as various collections of libraries and dictionaries, so it's probably not a good idea to build similar apps since there'll be some significant competition.

We also notice there are quite a few apps built around the book Quran, which suggests that building an app around a popular book can be profitable. It seems that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets.

However, it looks like the market is already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.


## Conclusions

In this project, we analyzed data about the App Store and Google Play mobile apps with the goal of recommending an app profile that can be profitable for both markets.

We concluded that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets. The markets are already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.